File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/ListeningHistory.cs
namespace IteratorDesignPattern.Example3;

public class ListeningHistory : IAggregate<Song>
{
    private readonly Stack<Song> _history = new();

    public void AddToHistory(Song song)
    {
        _history.Push(song);
    }

    public Stack<Song> GetHistory() => _history;

    public IIterator<Song> CreateIterator()
    {
        return new HistoryIterator(this);
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/src/main/java/behavioral/iterator/example3/ListeningHistory.java
package behavioral.iterator.example3;

import java.util.*;

public class ListeningHistory implements IAggregate<Song> {
    private final Deque<Song> history = new ArrayDeque<>();

    public void addToHistory(Song song) { history.push(song); }
    public Deque<Song> getHistory() { return history; }

    @Override
    public IIterator<Song> createIterator() { return new HistoryIterator(this); }
}
